modules/service/src/main/scala/lucuma/odb/service/AttachmentService.scala

```diff
@@ -0,0 +1,358 @@
+      )(fn =>
+        if (path.names.length > 1) {
+          InvalidName("File name cannot include a path").asLeft
+        }
+        else FileName(fn).asRight
+      )
+    }
+
+    extension (fileName: FileName) 
+      def extName: Option[NonEmptyString] = 
+        NonEmptyString.from(Path(fileName.value.value).extName).toOption
+  }
+
+  def fromConfigAndSession[F[_]: Async: Trace](
+    awsConfig: Config.Aws,
+    session:   Session[F]
+  ): Resource[F, AttachmentService[F]] = {
+    // We can switch back to unsafeFrom when a new version of refined is out that fixes 
+    // https://github.com/fthomas/refined/issues/1161 or a version of fs2-aws > 6.0.0
+    // is availabile with my PR that changes PartSizeMB to "GreaterEqual[5]"
+    // val partSize = PartSizeMB.unsafeFrom(5)
+    val partSize = refineV[Greater[5] Or Equal[5]](5).toOption.get
+    
+    val credentials = AwsBasicCredentials.create(awsConfig.accessKey.value, awsConfig.secretKey.value)
+
+    val s3R: Resource[F, S3AsyncClientOp[F]] = Interpreter[F].S3AsyncClientOpResource(
+      S3AsyncClient
+        .builder()
+        .credentialsProvider(StaticCredentialsProvider.create(credentials))
+        .serviceConfiguration(
+          S3Configuration
+            .builder()
+            .pathStyleAccessEnabled(true)
+            .build()
+        )
+        .region(Region.US_EAST_1)
+    )
+
+    s3R.map(s => (s, S3.create[F](s))).map{ (s3Ops, s3) =>
+
+      // We also use this to make sure the file exists and we have read permissions.
+      def getRemoteFileSize(fileKey: FileKey): F[Long] =
+        Trace[F].span("getRemoteFileSize") {
+          s3Ops
+            .headObject(
+              HeadObjectRequest
+                .builder()
+                .bucket(awsConfig.bucketName.value.value)
+                .key(fileKey.value.value)
+                .build
+            )
+            .onError { case e => Trace[F].attachError(e, ("error", true)) }
+            .map(_.contentLength())
+        }
+
+      def uploadRemoteFile(fileKey: FileKey, data: Stream[F, Byte]): F[Long] =
+        Trace[F].span("uploadRemoteFile") {
+          val f = for {
+            ref <- Ref.of(0L)
+            pipe = s3.uploadFileMultipart(awsConfig.bucketName, fileKey, partSize)
+            _    <- data.evalTapChunk(c => ref.update(_ + 1)).through(pipe).compile.drain
+            size <- ref.get
+          } yield size
+          f.onError { case e => Trace[F].attachError(e, ("error", true)) }
+        }
+
+      // TODO: eventually will probably want to check for write access for uploading/deleting files.
+      def withAccess[A](user: User, programId: Program.Id)(f: F[A]): F[A] = user match {
+        // guest users not allowed to upload files - at least for now.
+        case GuestUser(_) => Async[F].raiseError(AttachmentException.Forbidden)
+        case _            =>
+          ProgramService.fromSessionAndUser(session, user)
+            .userHasAccess(programId)
```

Comment:
This is going to end up running in a different transaction on the session, so you probably want to internalize the access check and run your operations within a transaction.

Reply:
:+1: Yeah the second case we probably don't want to hold the transaction open. If access is revoked during the upload they'll get a freebie but I think that's a case we don't need to worry about.

Reply:
We could always check again after the upload, if so inclined.  I agree not worth worrying about.